Attribute Base Access Control core functions

diegomasini opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

We have been reworking PR #1513 with @andresgaragiola and @raullaprida to split it in smaller PRs. This is the first one and it includes the core functions needed for Attribute Base Access Control. 
## Description

<!--- Describe your changes in detail. -->

Core functions for Attribute Based Access Control needed by Membership Services, the client and chaincode were factored out in a common package that can be included in any of the three components mentioned. Importing this package does not provide any dependency with the rest of the fabric, just with the utils and the primitives packages. These three packages with common functionality could be extracted in a separated library.
## Motivation and Context

This is a required feature to address #1139
## How Has This Been Tested?

Current behave tests has been run without errors.
Current unit tests has been run without errors.
Additional unit tests where included (attributes_test.go in core/crypto/abac).
A new behave test is pending we will work with @jeffgarratt on that.
## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff)
- [x] Either no new documentation is required by this change, OR I added new documentation
- [x] Either no new tests are required by this change, OR I added new tests
- [x] I have run [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports), [go vet](https://golang.org/cmd/vet/), and [golint](https://github.com/golang/lint). I have cleaned up all valid errors and warnings in code I have added or modified. These tools may generate false positives. Don't be worried about ignoring some errors or warnings. The goal is clean, consistent, and readable code.

Signed-off-by: Diego Masini damasini@us.ibm.com / diego.masini@gmail.com

Commit:
Revert "Adding read attributes AI"

This reverts commit 4695f6d9dcf317139a97b221e3e87a1296b0d81f.

diegomasini committed:
Merge remote-tracking branch 'upstream/master'

diegomasini committed:
Merge remote-tracking branch 'upstream/master'

diegomasini committed:
Merge remote-tracking branch 'upstream/master'

diegomasini committed:
Merge with upstream

diegomasini committed:
Merge remote-tracking branch 'upstream/master'

diegomasini committed:
Added missing as1 Object Identifier TCertAttributesHeader

diegomasini committed:
Merge remote-tracking branch 'upstream/master'

diegomasini committed:
Merge remote-tracking branch 'upstream/master'

diegomasini committed:
Merge with upstream

diegomasini committed:
Added abac common functions and initial unit tests

diegomasini committed:
Added abac common functions and initial unit tests

Added abac unit tests and resources

Fixed loadTCertAndPreK0 function to properly load the test tcert

Added check for negative positions when reading attributes by position

Removed unnecesary error returns

andresgaragiola committed:
Merge remote-tracking branch 'diego/abac-core' into abac-core

diegomasini committed:
Added abac unit tests and resources

diegomasini committed:
Fixed loadTCertAndPreK0 function to properly load the test tcert

diegomasini committed:
Added check for negative positions when reading attributes by position

srderson commented on core/crypto/abac/attributes.go:
```diff
@@ -0,0 +1,271 @@
+	return value, err
+}
+
+func createABACHeaderEntry(preK0 []byte) (*pb.ABACMetadataEntry, error) {
```
Should `error` be part of the return definition if the function never returns an error?

diegomasini replied:
Yes, the error comes from the getKAndValueForAttribute call, it could be an error or nil. Are you thinking about a different way to achieve this?

srderson replied:
My comment is for the `createABACHeaderEntry` function. It never calls `getKAndValueForAttribute`

srderson commented on core/crypto/abac/attributes.go:
```diff
@@ -0,0 +1,271 @@
+	return &pb.ABACMetadataEntry{HeaderAttributeName, attKey}, nil
+}
+
+func createABACMetadataEntry(attributeName string, preK0 []byte) (*pb.ABACMetadataEntry, error) {
```
Should `error` be part of the return definition if the function never returns an error?

diegomasini replied:
Yes, the error comes from the getKAndValueForAttribute call, it could be an error or nil. Are you thinking about a different way to achieve this?

srderson replied:
My comment is for the `createABACMetadataEntry` function. It never calls `getKAndValueForAttribute`

srderson commented on core/crypto/abac/attributes.go:
```diff
@@ -0,0 +1,271 @@
+	if err == nil {
+		entries = append(entries, headerEntry)
+	}
+	return &pb.ABACMetadata{metadata, entries}, nil
```
Should any of the `err` variables be returned here?

diegomasini replied:
You are right, let me fix it

srderson commented:
Thanks @diegomasini, much easier to review a commit of this size. You probably want to squash your commits as there are so many for this change.

diegomasini committed:
Merge remote-tracking branch 'upstream/master'

diegomasini committed:
Merged with remote

diegomasini commented:
@srderson I just removed the err variables from CreateABACMetadataObjectFromCert() you mentioned and I squashed the changes of this PR.

srderson commented:
@adecaro please give this a quick review when you have a chance. Thanks

adecaro commented on core/crypto/abac/attributes.go:
```diff
@@ -0,0 +1,267 @@
+	// TCertAttributesHeaders is the ASN1 object identifier of attributes header.
+	TCertAttributesHeaders = asn1.ObjectIdentifier{1, 2, 3, 4, 5, 6, 9}
+
+	padding = []byte{255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255}
```
What is the purpose of padding and why does it have that specific value?

diegomasini replied:
When a value is decrypted the decryption does not fail if someone uses the wrong key. It just produce garbage information. The padding was introduced to allow someone to verify that the decryption was successful. It needs to be a constant known value.

adecaro commented on core/crypto/abac/attributes.go:
```diff
@@ -0,0 +1,267 @@
+//EncryptAttributeValue encrypts "attributeValue" using "attributeKey"
+func EncryptAttributeValue(attributeKey []byte, attributeValue []byte) ([]byte, error) {
+	value := append(attributeValue, padding...)
+	return primitives.CBCPKCS7Encrypt(attributeKey, value)
```
Because the attributes are used in multiple places, shall we use GCM mode to protect against malleability attacks? The primitives package offers such an encryption mode.

diegomasini replied:
Attribute's keys are distributed in the transaction by using the provided confidentiality mechanism. I would like to review this with you since for the asset transfer use case we need to use GCM and potentially modify the transaction structure. But for the scope of this PR I think the confidentiality mechanism is enough. Also, I don't know is this is clear, but attributes are encrypted each time with a different key.

adecaro commented on core/crypto/abac/attributes.go:
```diff
@@ -0,0 +1,267 @@
+
+//getAttributeKey returns the attributeKey derived from the preK0 to the attributeName.
+func getAttributeKey(preK0 []byte, attributeName string) []byte {
+	return primitives.HMACTruncated(preK0, []byte(attributeName), 32)
```
Shall we add a prefix to []byte(attributeName) in order to distinguish the scope of the keys derived from preK0? 

Suppose that in the future we will use preK0 to derive keys wth different scopes. A way to identify the scope is by adding a prefix (associated to that specific scope) to the argument. This offers more modularity and retro compatibility. Of course, if we are pretty sure that preK0 will be used only for one thing, then we don't need this prefixing :)

diegomasini replied:
Attributes are associated to a given affiliation. In order to request a batch of TCerts the client needs to provide its ECert. That way attributes are implicitly scoped to that affiliation. It is possible to add a prefix to the attribute name if needed.

adecaro commented on core/crypto/abac/proto/abac.proto:
```diff
@@ -0,0 +1,33 @@
+    bytes AttributeKey = 2;
+}
+
+//ABACMetadata hold both the original metadata bytes and the metadata required by the Attribute Based Access Control
```
Just a clarification: Is the marshaled version of ABACMetadata what we will have to put in the transaction metadata for access control?

diegomasini replied:
Yes, the use is part of the client. And there is also an example of its use but is part of a coming PR as part of the split of the initial PR.

adecaro commented on core/crypto/abac/attributes.go:
```diff
@@ -0,0 +1,267 @@
+	var err error
+	var headerRaw []byte
+	encrypted := false
+	if headerRaw, err = utils.GetCriticalExtension(tcert, TCertAttributesHeaders); err != nil {
```
Just a clarification: Does the function assume that the x509 certificate has been already validated? I guess so, given that certificates must be always validated before extracting any information from them.

diegomasini replied:
Yes, at this point the TCert is validated in the stub, that code is not part of this PR since we are splitting a bigger PR in smaller pieces.

adecaro commented:
Hi @diegomasini, a way to avoid to have all these "Merge remote-tracking branch 'upstream/master'" is to rebase rather then merging. This helps reducing the number of commits :)

andresgaragiola committed:
Merge remote-tracking branch 'diego/abac-core' into abac-core

christo4ferris commented on core/crypto/abac/attributes_test.go:
```diff
@@ -0,0 +1,229 @@
+		t.Error(err)
+	}
+
+	_, attribute, err := getKAndValueForAttribute("position", prek0, tcert)
```
why are we (only) testing private functions? Seems to me that we should be testing both, but most importantly the public functions. Also, we only have positive tests here. Should have tests that ensure the correct processing of calls that pass incorrect argument values.

andresgaragiola replied:
TestGetKAndValueForAttribute is a test for private function "getKAndValueForAttribute", this private function is called by two public functions "GetValueForAttribute" and "GetKForAttribute". "GetValueForAttribute" is tested in the function TestGetValueForAttribute and we are adding a new test now called "TestGetKForAttribute" to test GetKForAttribute. Other public functions tests currently exists in this file. TestReadTCertAttributeByPosition_InvalidPositions is an example of a negative case test we will be adding more negative cases tests to this file.

christo4ferris commented on core/crypto/abac/attributes_test.go:
```diff
@@ -0,0 +1,229 @@
+		t.Error(err)
+	}
+
+	value, err := GetValueForAttribute("position", prek0, tcert)
```
should include cases for incorrect arguments and validate expected errors are returned, etc

christo4ferris commented on core/crypto/abac/attributes_test.go:
```diff
@@ -0,0 +1,229 @@
+		t.Error(err)
+	}
+
+	encryptedAttribute, err := ReadTCertAttributeByPosition(tcert, 1)
```
we should really be including more variability in the test cases. What if the function could only return the 1st position? please add more test cases.

diegomasini replied:
I think I don't understand completely here, position is an attribute holding the position inside the company, not a position inside the TCert.

christo4ferris replied:
to clarify my comment: if the function only returned the cert at position 1, this test passes and everything looks fine... but requesting the cert at position 2 would fail - and clearly that would be a bug.

christo4ferris commented on core/crypto/abac/attributes_test.go:
```diff
@@ -0,0 +1,229 @@
+	attributes["company"] = 1
+	attributes["position"] = 2
+
+	header := string(BuildAttributesHeader(attributes)[:])
```
this public function is only being indirectly tested, should have tests that check for things like passing an empty set of attributes, for instance.

christo4ferris commented on core/crypto/abac/attributes.go:
```diff
@@ -0,0 +1,267 @@
+}
+
+//GetABACMetadata object from the original metadata "metadata".
+func GetABACMetadata(metadata []byte) (*pb.ABACMetadata, error) {
```
Please add tests for this public function

christo4ferris commented on core/crypto/abac/attributes.go:
```diff
@@ -0,0 +1,267 @@
+}
+
+//CreateABACMetadata create the ABACMetadata from the original metadata
+func CreateABACMetadata(raw []byte, metadata []byte, preK0 []byte, attributeKeys []string) ([]byte, error) {
```
Please add tests for this

christo4ferris commented on core/crypto/abac/attributes.go:
```diff
@@ -0,0 +1,267 @@
+	return &pb.ABACMetadata{metadata, entries}
+}
+
+//CreateABACMetadataFromCert creates the ABACMetadata from the original metadata and certificate "cert".
```
please add tests for this

christo4ferris commented on core/crypto/abac/attributes.go:
```diff
@@ -0,0 +1,267 @@
+}
+
+//CreateABACMetadataObjectFromCert creates an ABACMetadata object from certificate "cert", metadata and the attributes keys.
+func CreateABACMetadataObjectFromCert(cert *x509.Certificate, metadata []byte, preK0 []byte, attributeKeys []string) *pb.ABACMetadata {
```
please add tests for this

andresgaragiola committed:
Added TestGetKForAttribute to test GetKForAttribute.

christo4ferris commented:
test coverage is currently 67% for this new component, please use a tool such as go-carpet to flesh out the remaining unit test coverage.

andresgaragiola committed:
Merge remote-tracking branch 'origin/abac-core' into abac-core

andresgaragiola committed:
Added several unit tests to abac library.

diegomasini committed:
Added more unit tests to increment coverage for new abac feature

diegomasini committed:
Merge remote-tracking branch 'upstream/master' into abac-core

diegomasini commented:
@srderson @binhn could you take a look at this PR again? We addressed the comments posted here and as @christo4ferris suggested we are using go-carpet to measure the testing coverage. With the last unit tests it is now 86.2% Remember that this PR needs to be processed to allow us to send the rest of the PRs that constitutes the whole ABAC feature.

christo4ferris commented:
86% is decent thank you... I can't review in detail now but am pleased with the coverage.

binhn commented:
@diegomasini Thanks Diego for your patience and collaboration on this pull request. 

I looked over this again tonight, and one last question: In `abac.proto`, you define `ABACMetadata`. Why metadata? The content obviously contains an array of attributes, not something that we don't know. Also `ABACMetadata` contains `Metadata` as bytes. It is not clear to me why we need both `Metadata` and `Entries` in this struct. Could you offer some explanation as part of comment of the struct.

andresgaragiola commented:
thanks for your comment @binhn. ABACMetadata includes both the original metadata and one entry per each attribute included in the transaction. Each ABACMetadataEntry includes the attributeName and the key required to decrypt that attribute. We have added the metadata field to wrap the existing metadata as it is used currently. We also provided an API to read the attributes from it.

binhn commented:
@diegomasini @andresgaragiola  As discussed, we agreed to change the package name, including files and functions, so that it would be more general since attributes are not only for abac but can be used anyhow the app wishes. We also agreed to defer the resolution to the problem of having the peer knows about the decryption key of the attribute keys, until we complete other parts of attributes so that other work may continue.

andresgaragiola committed:
Modified library name from 'abac' to 'attributes' and added more tests reaching 92.2% of coverage.

andresgaragiola committed:
Merge remote-tracking branch 'diego/abac-core' into abac-core

andresgaragiola committed:
Merge remote-tracking branch 'upstream/master' into abac-core

andresgaragiola commented:
@binhn as we discussed today, we recently have pushed a new version of this library renamed to 'attributes' instead of 'abac' and some additional test increasing the coverage to 92%.

binhn commented:
@andresgaragiola Excellent.  Thank you!
